### PR TITLE
release/2.0

### PR DESCRIPTION
**Pull request by DarkGhostHunter**
[CLA Signed]

This pull is a remake of the ReCaptcha package. Notable changes are:

* Namespace change.
* Simpler verification (one liner)
* Methods better reflect the response from reCAPTCHA servers.
* Removed useless classes.
* PSR-18 client compatible, HTTP/2 with cURL as suggested.
* Reworked README.
* Added UPGRADING.
* Removed unsupported PHP versions (< 7.1)
* Tests with high code coverage

You can check out the [README](https://github.com/google/recaptcha/blob/ef1d1560ec3dfa3e5890bb373cd562b89ec3ee28/README.md).

### CODE REVIEW

**Comment by DarkGhostHunter** [comment]
Got bored and completed all tests. This pull is pending revision unless further

**Comment by DarkGhostHunter** [comment]
While this package comes with very barebones HTTP Clients, I think it would be better to avide to PSR-18 and let the developer include it's own HTTP Client, but including a default.

Instead of using multiple clients (Curl, Stream, Socket), I think it should require Symfony HTTP Client. Considering this requires latest version of PHP, and installing/using Composer is not that difficult, I'm gonna make the transition to it instead of relying on the clients.

Yeah, goodbye clients, hello PSR-18. The older version should be left unsupported.

**Comment by DarkGhostHunter** [comment]
3:00 AM "Just one more line of code".

This commit uses PSR-18 HTTP Clients and PSR-17 factories. It **doesn't include them**, but these are suggested in the README. This way developers can have full control on how their servers contact reCAPTCHA servers (like using Guzzle along with their framework own PSR-17 factories already installed, avoiding code redundancy).

I went for Symfony 5.0 since it's mature and PSR-18 compatible out of the box.

**Comment by DarkGhostHunter** [comment]
Any signs of life for this project would be appreciated.

**Comment by rowan-m** [comment]
@DarkGhostHunter Hey there, this PR is a **lot** to process. A major version change with BC breaks, new dependencies, etc. needs careful thought and review. I will take the time to look at this and provide feedback.

**Comment by DarkGhostHunter** [comment]
> @DarkGhostHunter Hey there, this PR is a **lot** to process. A major version change with BC breaks, new dependencies, etc. needs careful thought and review. I will take the time to look at this and provide feedback.

No problem. Anything, hit the comments.

**Comment by DarkGhostHunter** [comment]
Almost two months have passed. Any updates?

**Comment by mkarnicki** [comment]
I'm usually the one to +1 via reaction and not add an unnecessary comment, but this work by @DarkGhostHunter deserves your attention @rowan-m .

File: src/ReCaptcha.php [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+     */
+    public function getClient()
+    {
+        return $this->client;
+    }
+
+    /**
+     * Sets the HTTP Client to use with this.
+     *
+     * @param  mixed $client The object that this client will use for
```
Comment (williamdes) [inline src/ReCaptcha.php]: Both getter, property use the interface
For me you should replace mixed by the interface

File: src/ReCaptcha.php [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+        $this->secret = $secret;
+
+        return $this;
+    }
+
+    /**
+     * Verifies a challenge and returns a response.
+     *
+     * @param  string $token
+     * @param  string|null $ip
```
Comment (williamdes) [inline src/ReCaptcha.php]: I am not okay with this phpdoc
In strict mode if you give a null it will fail

File: src/ReCaptcha.php [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+        $response = $this->parseResponse($array);
+
+        return $this->checkErrors($response);
+    }
+
+    /**
+     * Receives a request and returns a response from reCAPTCHA servers.
+     *
+     * @param  string $token The token that identifies the reCAPTCHA challenge.
+     * @param  string|null $ip The optional IP of the user challenged.
```
Comment (williamdes) [inline src/ReCaptcha.php]: Same for each occurrence (or I am wrong?)

File: src/ReCaptcha.php [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+     * Quickly validate a reCAPTCHA challenge
+     *
+     * @codeCoverageIgnore
+     * @param  string $secret
+     * @param  string $token
+     * @param  string|null $ip
+     * @return bool
+     * @throws \Psr\Http\Client\ClientExceptionInterface
+     */
+    public static function validate(string $secret, string $token, string $ip = null)
```
Comment (williamdes) [inline src/ReCaptcha.php]: You could use return type hints because you require php 7

File: src/ReCaptchaResponse.php [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+     */
+    public function invalid()
+    {
+        return ! $this->valid();
+    }
+
+    /**
+     * Return if the response has a particular error
+     *
+     * @param  mixed ...$errors
```
Comment (williamdes) [inline src/ReCaptchaResponse.php]: array

File: src/ReCaptchaResponse.php [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+    public function constraints()
+    {
+        return $this->constraints;
+    }
+
+    /**
+     * Sets the constraints to check this response.
+     *
+     * @param  array $constraints
+     * @return $this
```
Comment (williamdes) [inline src/ReCaptchaResponse.php]: `self` seemed a better practice

File: src/ReCaptchaResponse.php [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+    {
+        $this->constraints = $constraints;
+
+        return $this;
+    }
+
+    /**
+     * Dynamically return an attribute as a property.
+     *
+     * @param $name
```
Comment (williamdes) [inline src/ReCaptchaResponse.php]:
```suggestion
     * @param string $name The attribute name
```

File: composer.json [hunk 8]
```diff
@@ -2,38 +2,61 @@
...
+        "ext-json": "*",
+        "psr/http-client": "1.*",
+        "psr/http-factory": "1.*"
     },
     "require-dev": {
-        "phpunit/phpunit": "^4.8.36|^5.7.27|^6.59|^7.5.11",
-        "friendsofphp/php-cs-fixer": "^2.2.20|^2.15",
+        "nyholm/psr7" : "1.*",
+        "symfony/http-client": "5.*",
+        "phpunit/phpunit": "^8.0",
```
Comment (williamdes) [inline composer.json]:
```suggestion
        "phpunit/phpunit": "^8 | 9^",
```

File: src/ReCaptchaResponse.php (hunk 6 above)
Comment (DarkGhostHunter) [reply src/ReCaptchaResponse.php]: I usually use this:
* `$this` if its the same instance for a dynamic method (like fluent/chained methods).
* `self` only for interfaces.
* `static` for static methods or when returning a new instance.

Also, there is no PHPDoc documentation of what on when, AFAIK.

Comment (williamdes) [reply src/ReCaptchaResponse.php]: https://docs.phpdoc.org/latest/references/phpdoc/tags/return.html
"When provided it MUST contain a Type to indicate what is returned"
https://docs.phpdoc.org/latest/references/phpdoc/types.html
```
keyword                  = "string"|"integer"|"int"|"boolean"|"bool"|"float"
                           |"double"|"object"|"mixed"|"array"|"resource"|"scalar"
                           |"void"|"null"|"callback"|"false"|"true"|"self"
```

So maybe replace $this ?

File: tests/ReCaptchaTest.php [hunk 9]
```diff
@@ -0,0 +1,492 @@
...
+    {
+        $action = '/unsanitized@action/to-test_here?foo=bar&quz=qux';
+
+        $recaptcha = ReCaptcha::make('test_secret')
+            ->saneAction($action);
+
+        $this->assertEquals('/unsanitizedaction/to_test_here', $recaptcha->getConstraints()['action']);
+    }
+
+    public function testConstraintsErrors()
```
Comment (williamdes) [inline tests/ReCaptchaTest.php]:
```suggestion
    public function testConstraintsErrors(): void
```

could be used on all tests

Comment (DarkGhostHunter) [reply tests/ReCaptchaTest.php]: I don't think it's really necessary unless somebody would extend on it or there performance reasons behind it.

Comment (DarkGhostHunter) [reply tests/ReCaptchaTest.php]: I don't think it's really necessary unless somebody would extend on it or there performance reasons behind it.

Comment (williamdes) [reply tests/ReCaptchaTest.php]: As you want :)

**Comment by DarkGhostHunter** [comment]
image

Gulge pls

**Comment by DarkGhostHunter** [comment]
I'm still waiting for any useful response.

**Comment by DarkGhostHunter** [comment]
Seems that has been total radio silence. This is not a good sign. I'm closing this.

**Pull request closed without merging**